Add filament lookup plugin

HD650 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Add a filament lookup plugin to show the content of process table.
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
filament_lookup
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (filament-lookup 8a59823ae5) last updated 2018/06/05 17:31:41 (GMT -400)
  config file = None
  configured module search path = [u'/home/zhikangzhang/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/zhikangzhang/workspace/github/hd650/ansible/lib/ansible
  executable location = /home/zhikangzhang/workspace/github/hd650/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```

Commit:
Add filament lookup plugin

Filament lookup plugin show the content of process table.

Commit:
unittest infrastructure of filament lookup

Prior to this change,

This change

Commit:
Unittest now is functional

Commit:
Fix sanity issues

Commit:
Add integration test for lookup plugin

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+from ansible.plugins.lookup import LookupBase
+
+try:
+    import subprocess
```
Are you sure you want a fallback display if this import statement fails?

HD650 replied:
Oh, this will cause fallback display? I just see other plugins do this and I copied. What's the right way to import display?

webknjaz replied:
The problem I'm trying to spot is that you have wrapped a number of imports here, so if **any** of them fails:
1) next (import) lines in try block **won't be executed at all**
2) control flow will jump into except block, so that you'll have other import which will be executed and then `display` var initialized

After that, all other code in module will get executed and it seems to rely on the fact that all of other imports succeeded.
In case if they failed you'll get a `NameError`, which will state that it doesn't know about `AnsibleError`, for example, and it doesn't exist.

That workaround is designed to provide a fallback for `display` if importing it from `__main__` module fails.
But by wrapping all the imports you're simply masking lots of different errors, which would have to be processed separately.

So the rule of thumb would be to narrow down try block to some concrete piece of code, which you expect to raise an exception. It is considered harmful to "swallow" arbitrary errors without proper handling of corresponding situations.

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+def run_command(terms):
+    command = None
+    # if no argument passed, show whole process table
+    if len(terms) is 0:
```
This is wrong. `is` operator is an identity comparison.

Under the hood it's `id(len(terms)) == id(0)`, which effectively compares the addresses of two integers in memory in this case.
It only works here because cpython caches some amount of integers for later reuse.
Proper use of it would be `len(terms) == 0`, which loosely corresponds to `len(terms).__eq__(0)`. Or you could do `if not len(terms)`, which would work because `not 0` is `True` (because `bool(0) is False`).

So the rule of thumb is to always use `is` for comparison with `None`, `True` or `False`. In other cases use numeric comparisons (`==`, `!=`, `>`, `<=` etc.), unless specifically needed.

https://docs.python.org/3/reference/datamodel.html
https://stackoverflow.com/a/133024/595220

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+    if len(terms) is 0:
+        command = "ps aux"
+    # if get an argument, search it in the process table
+    elif len(terms) is 1:
```
same

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+        command = "ps aux"
+    # if get an argument, search it in the process table
+    elif len(terms) is 1:
+        command = "ps aux|grep " + str(terms[0])
```
Why do you accept a list and then extract a single arg from there anyway? Maybe you could just concatenate all items there?

webknjaz replied:
Oh, and in Python we tend to avoid string concatenation with "+"

HD650 replied:
It's because the lookup plugin has no arguments limitation. So the argument the plugin received is already a list. And I put the argument validation code in run_command because it's easy for unit test. So the right way is to have a function do the validation and a function run the command?

webknjaz replied:
You invoke it yourself, so do validation in lookup plugin class

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+    elif len(terms) is 1:
+        command = "ps aux|grep " + str(terms[0])
+    # other condition, raise exception
+    else:
```
You'd probably want to do validation as a first thing in the function and then stop caring about checking things. This seems a logical thing to do.

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+
+
+def run_command(terms):
+    command = None
```
I would probably go for an immediate default:
```python
command = "ps aux"
if terms:
  command = "%s | grep %s" % (command, ' '.join(terms))
```

Commit:
Fix the import fallback problem

Prior to this change, if some import statement failed in filament lookup
plugin, the statements below will not be executed.

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+    display = Display()
+
+
+def run_command(terms):
```
You could probably use argument unpacking if you tend to accept multiple args:
```python
def run_command(*terms):
    command = 'ps aux'
    if terms:
        command = '%s | grep %s' % (command, ' '.join(terms))

```

https://docs.python.org/3/tutorial/controlflow.html#more-on-defining-functions

Also common way of having optional function args is to default them to `None`:
```python
def run_command(terms=None):
    if terms is None:
        terms = []  # have some reasonable default if you need
```

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+    display.vvv(command)
+
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    result = p.stdout.read()
```
I bet it's possible that this thing raises `IOError`

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+        raise AnsibleError("Argument Fault: 1 string argument expect, {0} got.".format(len(terms)))
+    display.vvv(command)
+
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
I guess this might raise an `OSError`

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,62 @@
+
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    result = p.stdout.read()
+    return to_text(result)
```
So refactored I would have a function like this, in case if you want to have **one optional arg for grep**:
```python
from ansible.module_utils import six
...
...
def run_command(needle=None):
    command = "ps aux"
    if needle is not None:
        command = '%s | grep "%s"' % (command, needle)
    display.vvv(command)

    try:
        proc = (
            subprocess.
            Popen(
                command,
                shell=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
            )
        )
        return to_text(proc.stdout.read())
    except OSError as os_err:
        six.raise_from(AnsibleError("Failed to fetch processes."), os_err)
    except IOError as io_err:
        six.raise_from(AnsibleError("Failed to read subprocess output."), io_err)
```

webknjaz commented on test/units/plugins/lookup/test_filament_lookup.py:
```diff
@@ -0,0 +1,72 @@
+from ansible.compat.tests import unittest
+from ansible.compat.tests.mock import MagicMock, patch
+import ansible.plugins.lookup.filament_lookup as filament_lookup
+import subprocess
```
Please note that it's a good practice to separate stdlib imports from own project imports and from third-party lib imports.

The order should be as follows:
1. stdlib imports
2. empty line
3. third-party
4. empty line
5. own

HD650 replied:
Ok, and I just want to reply your review of fallback import and it disappeared. So once somebody changes the code, the corresponding review will disappear?

webknjaz replied:
It collapses, but still accessible from the conversation tab.

webknjaz replied:
https://github.com/ansible/ansible/pull/41161#pullrequestreview-126577732 it's here, just click "Show outdated" next to it.

webknjaz replied:
This thing, right? https://github.com/ansible/ansible/pull/41161#discussion_r193729928

HD650 replied:
Yes, I saw it.

webknjaz commented on test/units/plugins/lookup/test_filament_lookup.py:
```diff
@@ -0,0 +1,72 @@
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
```
usually, we don't wrap imports with parentheses unless we need multiline

webknjaz commented on test/units/plugins/lookup/test_filament_lookup.py:
```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+# (c) 2015, Toshio Kuratomi <tkuratomi@ansible.com>
```
@HD650 don't copy-paste blindly

HD650 replied:
...., Yes, next time I will read it first.

Commit:
Adjust the order of import

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -36,13 +36,13 @@
 
 def run_command(terms):
-    command = None
+    command = "ps aux"
```
Please don't mix unrelated changes in one commit. You may use `git commit --patch` for this.

Commit:
Refactored lookup plugin

webknjaz commented on lib/ansible/plugins/lookup/filament_lookup.py:
```diff
@@ -0,0 +1,66 @@
+        result = p.stdout.read()
+    except OSError as os_err:
+        six.raise_from(AnsibleError("Failed to fetch processes."), os_err)
+    except ISError as io_err:
```
`OSError`

webknjaz replied:
`IOError`

webknjaz commented on test/integration/targets/filament_lookup/tasks/main.yml:
```diff
@@ -0,0 +1,6 @@
+- name: integration test for one arg lookup
+  assert:
+    that: " one_arg_lookup is defined and 'grep' in one_arg_lookup "
```
```yaml
  that: >
    one_arg_lookup is defined and 'grep' in one_arg_lookup
```

webknjaz replied:
https://learnxinyminutes.com/docs/yaml/

webknjaz replied:
https://stackoverflow.com/a/21699210/595220

webknjaz commented on test/units/plugins/lookup/test_filament_lookup.py:
```diff
@@ -0,0 +1,72 @@
+
+from ansible.errors import AnsibleError
+from ansible.compat.tests import unittest
+from ansible.compat.tests.mock import MagicMock, patch
```
https://pypi.org/project/pytest-mock/

Commit:
Chang the author info

Commit:
Adjust the order of import

Commit:
Fix the unittest

Commit:
Change doc

Commit:
Fix no module called epdb error

Commit:
Fix documents format error

Commit:
Fix a typo
